htmd/smallmol/chemlab/builder.py

```diff
@@ -0,0 +1,406 @@
+
+        self.smallmol = smallmol.copy()
+
+        if checkInitialConformer:
+            self._prepareConformer()
+
+    def _prepareConformer(self):
+        """
+        Generate a conformer and stores the coords in the SmallMol object
+        """
+
+        from rdkit.Chem import AllChem
+
+        sm = self.smallmol
+        mol = sm.toRdkitMol(includeConformer=True)
+
+        AllChem.EmbedMolecule(mol, AllChem.ETKDG())
+
+        conf = mol.GetConformer(0)
+        coords = conf.GetPositions()
+        coords = coords[:, :, np.newaxis]
+        sm.coords = coords
+
+
+    def addHydrogens(self, onlyExplicit=False):
+        """
+        Adds the hydrogens to fill up the valence of the atoms. With onlyExplicit is possible to add only the polar ones
+
+        Parameters
+        ----------
+        onlyExplicit: bool
+            If True, only the polar hydrogens will be added
+            Default: False
+        """
+
+        pT = self.periodictable
+        sm = self.smallmol
+
+        # polar hydrogens are the ones bonded to hetero atoms (no carbon)
+        if onlyExplicit:
+            heavy_sel = ' '.join(_hetero_atoms)
+        else:
+            heavy_sel = ' '.join(_heavy_atoms)
+
+        atomidxs = sm.get( 'idx', 'element {}'.format(heavy_sel))
+        formalcharges = sm.get('formalcharge', 'element {}'.format(heavy_sel))
+
+        missing_atoms = np.array([pT.getMissingValence(aId, sm, fch) for aId, fch in zip(atomidxs, formalcharges)])
+        ids = np.where(missing_atoms > 0)
+        heavyToFill = atomidxs[ids]
+        numHydrogens = missing_atoms[ids]
+        #print("missing hydrogens: ", numHydrogens)
+        h_atoms = []
+
+        for n_h, a in zip(numHydrogens, heavyToFill):
+            for i in range(n_h):
+                new_atom = pT.getAtom('H', a)
+                h_atoms.append(new_atom)
+
+        self._addAtoms(h_atoms)
+
+
+    def removeHydrogens(self, removePolars=True, removeNonPolars=True):
+        """
+        Removes the hydrogens. Is it possible to choose the type of hydrogens to remove: polar and not polar
+
+        Parameters
+        ----------
+        removePolars: bool
+            If True, the polar hydrogens will be removed
+            Default: True
+        removeNonPolars: bool
+            If True, the non polar hydrogens will be removed
+            Default: True
+        """
+
+        ## TODO
+        ## Fix the the kekulization problem whem removing hydrogen from NH in aromatic system.
+        sm = self.smallmol
+        hydrogensIdx =  sm.get('idx', 'element H')
+
+        neighborsIdx = np.concatenate(sm.neighbors[hydrogensIdx])
+        neighbors = sm.element[neighborsIdx]
+        hydrogensPolarIdx = hydrogensIdx[np.where( neighbors != 'C' )]
+        hydrogensNonPolarIdx = hydrogensIdx[np.where(neighbors == 'C')]
+
+
+        hydrogensIdx = np.array([],dtype=np.int)
+        if removePolars:
+            hydrogensIdx = np.append( hydrogensIdx, hydrogensPolarIdx)
```

Comment:
Don't append to numpy arrays (it's slow). Use lists and at the last step convert to np.array.

Also make sure you use the correct int dtype. Do you want `int32` or `int64` or even more correct `uint32`? Indexes cannot be negative and I doubt you will have more than 4billion atoms so I guess `uint32` is enough.

```python
hydrogensIdx = []
if bla:
   hydrogensIdx.append(x)
if blu:
   hydrogensIdx.append(y)
hydrogensIdx = np.array(hydrogensIdx , dtype=np.int32)
```

Reply:
I see. Due to the fact that I am using two numpy (in your example x and y are numpy arrays). I did like that:
```
hydrogensIdx = []
if removePolars:
   hydrogensIdx.extend(hydrogensPolarIdx.tolist())
if removeNonPolars:
   hydrogensIdx.extend(hydrogensNonPolarIdx.tolist())
hydrogensIdx = np.array(hydrogensIdx, dtype=np.uint32)
```

Reply:
Or you can append np.arrays to a list and at the end `np.vstack/hstack/stack` them. I usually do that. No converting to list and back to array would be ugly :D So either the way it was before with the correct dtype or as I said with stacking :)

Reply:
Hoping is correct, I modified with ```np.hstack``` like that:
```
hydrogensIdx = np.array([], dtype=int)
if removePolars:
   hydrogensIdx = np.hstack((hydrogensIdx, hydrogensPolarIdx))
if removeNonPolars:
   hydrogensIdx = np.hstack((hydrogensIdx, hydrogensNonPolarIdx))
```

Reply:
Wrong dtype though :D As I said `uint32` is more than enough. It cannot go negative so `int` which is `int64` is overkill (sorryyyyy)